# Aufgabe 2

In [1]:
import numpy as np

a = [[1.2969, 0.8648], [0.2161, 0.1441]]

b_1 = [0.8642, 0.1440]
b_2 = [0.86419999, 0.14400001]
b_3 = [0.8641999, 0.1440001]
b_4 = [0.864199, 0.144001]
b_5 = [0.86419, 0.14401]

b_s = [b_1, b_2, b_3, b_4, b_5]

for b in b_s:
    print(np.linalg.solve(a, b))

[ 2. -2.]
[ 0.9911 -0.487 ]
[-8.08900001 13.13000002]
[-98.89000015 149.30000023]
[-1006.90000156  1511.00000235]


Man sieht eine sehr starke Abweichung der Lösungen mit b2 bis b5 von der Lösung mit b1, wobei die Abweichung der Lösungen immer weiter zunimmt, je weiter die Eingangswerte von b1 abweichen.

Genauer:

Abweichung b2: 

$$
x: \frac{2 - 0.9911}{2} ≈ 0.504 = 50,4 \%
$$

$$
y: \frac{-2 - (-0.487)}{-2} ≈ 0,757 = 75,7 \%
$$

Abweichung b5:
$$
x: \frac{2 - (-1006.9)}{2} ≈ 504.45 = 50445 \%
$$

$$
y: \frac{-2 - 1511}{-2} ≈ 756,5 = 75650 \%
$$

# Aufgabe 3

$$
f(x) = \frac{tan(x) - x}{x^3}
$$

## Grenzwert x → 0

![image](limes_handwriting.png)

## Numerische Bestimmung mit immer kleineren x

In [1]:
import numpy as np

def f(x: float):
    return (np.tan(x) -x) / np.power(x, 3)

for k in range(1, 11):
    print(f(1 / np.power(10, k)))

0.33467208545054355
0.33334666720702394
0.33333346673158903
0.33333333651890806
0.3333328757329847
0.33330746474456724
0.3308722450212111
0.0
0.0
0.0


## Grund für Instabilität

Man sieht, dass für kleine x zunächst relativ genau der tatsächliche Grenzwert der Funktion f ermittelt werden kann. Doch bei immer kleiner werdenden x-Werten wird irgendwann der Punkt erreicht, bei dem mit dem numerischen Verfahren das Ergebnis 0 errechnet wird. Das hängt mit dem Aufbau der Funktion f und der Rundung zusammen. 

Sowohl tan(x) als auch x wird gegen null immer kleiner, sodass die Differenz ebenfalls immer kleiner wird. Beim numerischen Verfahren geht zwangsweise irgendwann, durch Rundung, der immer kleiner werdende Unterschied zwischen den beiden Werten verloren. Sobald diese nicht mehr unterschieden werden können, also wenn praktisch gilt: 

$$
tan(x) = x
$$

Ist die Differenz der Werte im Nenner null und somit der numerisch errechnete Grenzwert ebenfalls null.

Das führt zu einer Instabilität, weil ab einem bestimmten Punkt bei kleiner werdenden x-Werten plötzlich eine starke Abweichung von der tatsächlichen Lösung entsteht.